src/game_api/script/usertypes/state_lua.cpp

```diff
@@ -414,6 +415,31 @@ void register_usertypes(sol::state& lua)
     statememory_type["next_entity_uid"] = &StateMemory::next_entity_uid;
     statememory_type["room_owners"] = &StateMemory::room_owners;
 
+    auto state_get_user_data = [](StateMemory& state) -> sol::object
+    {
+        auto backend = LuaBackend::get_calling_backend();
+        auto local_datas = backend->local_state_datas;
+        if (local_datas.contains(&state))
+        {
+            return local_datas[&state].user_data;
+        }
+        return sol::nil;
+    };
+
+    auto state_set_user_data = [](StateMemory& state, sol::object user_data) -> void
+    {
+        auto backend = LuaBackend::get_calling_backend();
+        backend->local_state_datas[&state].user_data = user_data;
+    };
+    auto user_data = sol::property(state_get_user_data, state_set_user_data);
+
+    statememory_type["user_data"] = std::move(user_data);
+    /* StateMemory
+    // user_data
+    // You can store a table (or lua primitive) here and it will store data correctly in online multiplayer, by having a different copy on each state and being copied over when the game does.
+    // Doesn't support recursive tables / cyclic references. Metatables will be transferred by reference instead of being copied
```

Comment:
> Metatables will be transferred by reference instead of being copied

This is kind of obvious, since you just can't ever do a copy of metatable in lua

Reply:
metatables are just tables, with fields that are accessed by lua when the table is set as a metatable of one, so those are copiable technically, but copying them doesn't make sense for what they are used most, if not all of the time (which is mostly making OOP like methods for tables). Not sure about metatables created by sol2.
https://www.lua.org/pil/13.html